source/contents.js

```diff
@@ -1,34 +1,45 @@
-export const getStreamContents = async (stream, {init, convertChunk, getSize, addChunk, finalize}, {maxBuffer = Number.POSITIVE_INFINITY} = {}) => {
+export const getStreamContents = async (stream, {init, convertChunk, getSize, addChunk, getFinalChunk, finalize}, {maxBuffer = Number.POSITIVE_INFINITY} = {}) => {
 	if (!isAsyncIterable(stream)) {
 		throw new Error('The first argument must be a Readable, a ReadableStream, or an async iterable.');
 	}
 
-	let length = 0;
-	let contents = init();
-	const textDecoder = new TextDecoder();
+	const state = init();
+	state.length = 0;
 
 	try {
 		for await (const chunk of stream) {
 			const chunkType = getChunkType(chunk);
-			const convertedChunk = convertChunk[chunkType](chunk, textDecoder);
-			const chunkSize = getSize(convertedChunk);
-
-			if (length + chunkSize > maxBuffer) {
-				throw new MaxBufferError();
-			}
-
-			const newLength = length + chunkSize;
-			contents = addChunk(convertedChunk, contents, newLength, length);
-			length = newLength;
+			const convertedChunk = convertChunk[chunkType](chunk, state);
+			appendChunk({convertedChunk, state, getSize, addChunk, maxBuffer});
 		}
 
-		return finalize(contents, length, textDecoder);
+		appendFinalChunk({state, convertChunk, getSize, addChunk, getFinalChunk, maxBuffer});
```

Comment:
The final chunk (`�`) must use the same logic as other chunks, including the `maxBuffer` logic.
To do this required two changes:
  - Adding a `getFinalChunk()` method that returns either `undefined` (no final chunk) or the final chunk, with the right type
  - Extracting the `appendChunk()` logic so that it can be used after `getFinalChunk()`. To do this, the inside of the main `for` loop had to be extracted to its own function `appendChunk()`. To do this, stateful arguments had to be put into a `state` object. Changing this required refactoring some additional functions.